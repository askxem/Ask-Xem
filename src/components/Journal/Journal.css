.backdrop {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, 0.7);
  z-index: 1;
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal form {
  display: flex;
  flex-direction: column;
}

.modal button {
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
  border: none;
  background-color: #91ad8c;
  border-radius: 20px;
  padding: 10px;
  box-shadow: 0px 5px 10px gray;
  color: white;
}

.modal div {
  display: flex;
  gap: 15px;
}

.journal {
  border-radius: 10px;
  width: 75vw;
  height: 60vh;
  padding: 10px;
  text-overflow: scroll;
  background-color: white;
  font-family: 'Varela Round', sans-serif;
  color: black;
}
