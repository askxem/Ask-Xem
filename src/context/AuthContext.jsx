import { createContext, useState, useContext, useMemo } from "react";
import { getUser, signInUser, signOutUser, signUpUser } from "../services/users.js";


const AuthContext = createContext();

function ProvideAuth({children}) {

    const currentUser = getUser();

    const [user, setUser] = useState(
        currentUser ? {...currentUser} : {}
    );

    async function signUp(email, password) {
        const newUser = await signUpUser(email, password);
        setUser(newUser);
    }

    async function signIn(email, password) {
        const newUser = await signInUser(email, password);
        setUser(newUser);
    }

    async function signOut() {
        await signOutUser();
        localStorage.clear();
        setUser({});
        history.push('/home');
    }

    const value = useMemo(() => ({ user, signUp, signIn, signOut}), [user]);

    return (
        <AuthContext.Provider value={value} >
            {children}
        </AuthContext.Provider>
    )
}

/**
 * 
 * @returns a user object, a signup, signin, and signout fn that hits the database.
 */
function useAuth(){
    const context = useContext(AuthContext);

    if (context === undefined) throw new Error('Auth context not accessible outside of the auth provider.')

    return context;
}


export { ProvideAuth, useAuth }