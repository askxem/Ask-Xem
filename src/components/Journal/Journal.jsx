// Is going to be a modal. 
// https://www.youtube.com/watch?v=RF78mP1AMQ0


// need to implement a switch state which is true/false within the parent components to manage wether or not this is displayed.

import React, {useEffect, useState} from 'react';
import { Link } from 'react-router-dom';
import { motion, AnimatePresence } from 'framer-motion'
import styles from './Journal.css'
import { useAuth } from '../../context/AuthContext.jsx';
import { getJournal, insertJournal, updateJournal } from '../../services/journal.js';

const backdropVariants = {
    visible: {opacity: 1},
    hidden: {opacity: 0 }
}

export default function Journal({ showJournal, setShowJournal }) {
    const [journalInput, setJournalInput] = useState('')
    const {user} = useAuth();

    useEffect(() => {
        async function fetchJournal(){
            try{
                const resp = await getJournal(user.id);
                setJournalInput(resp.journal);
            } catch(e) {
                console.log(e);
                if (e.details === 'Results contain 0 rows, application/vnd.pgrst.object+json requires 1 row'){
                    const resp = await insertJournal(user.id, 'Welcome to your journal!');
                    setJournalInput(resp[0].journal);
                }
            }
        }
        
        fetchJournal();
    }, [])
    
    async function handleSubmit(e){
        e.preventDefault();
        await updateJournal(user.id, journalInput);
        setShowJournal(false);
    }

  return (
    //   allows us to animate the mounting and dismounting of nested components. 
    //  exitBeforeEnter allows all animations within the component to complete before it executes it's own.
      <AnimatePresence exitBeforeEnter>
          {
              showJournal && (
                <motion.div className={styles.backdrop} variants={backdropVariants} initial='hidden' animate='visible'>
                    {/* represents backdrop of modal. Will create the semi-transparent sheet that fades out background view.*/}
                    <motion.div className={styles.modal}>
                        <form onSubmit={handleSubmit}> 

                            <label htmlFor="journal">Journal</label>
                            <textarea 
                            id='journal' 
                            name='journal' 
                            value={journalInput} 
                            onChange={(e) => setJournalInput(e.target.value)} />

                            <button>Submit and Return to page.</button>
                        </form>
                    </motion.div>
                </motion.div>
              )}
      </AnimatePresence>
  )
}
