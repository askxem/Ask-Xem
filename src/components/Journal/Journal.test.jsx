import { render } from '@testing-library/react'
import { ProvideAuth } from '../../context/AuthContext.jsx';
import Journal from './Journal.jsx';

jest.mock("../../services/journal.js");
jest.mock('../../services/users.js');
jest.mock('../../context/AuthContext.jsx')

it('renders the journal component', () => {
    const { container } = render(
        <ProvideAuth mockUser={{ id: 2 }}>
            <Journal setShowJournal={jest.fn()} />
        </ProvideAuth>
    )

    expect(container).toMatchSnapshot();
})