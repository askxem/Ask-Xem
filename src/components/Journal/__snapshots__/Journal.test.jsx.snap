// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the journal component 1`] = `
<div>
  <div
    class="backdrop"
    style="opacity: 0;"
  >
    <div
      class="modal"
    >
      <form>
        <label
          for="journal"
        >
          Journal
        </label>
        <textarea
          id="journal"
          name="journal"
        />
        <button>
          Submit and Return to page.
        </button>
      </form>
    </div>
  </div>
</div>
`;
